File: 自写/Demo/ObjcRuntimeDemo/ObjcRuntimeDemo/Classes/KVO/KVOController.m
//
//  KVOController.m
//  ObjcRuntimeDemo
//
//  Created by luozhijun on 2017/3/16.
//  Copyright © 2017年 RickLuo. All rights reserved.
//

#import "KVOController.h"
#import "NSObject+KVO.h"

@interface KVOController ()
@property (strong, nonatomic) IBOutlet UITextField *textField;
@end

@implementation KVOController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(textFieldTextDidChange) name:UITextFieldTextDidChangeNotification object:nil];
}

- (IBAction)addObserver:(id)sender {
    [self.textField zj_addObserver:self forKeyPath:@"text" changed:^(NSObject *observedObj, NSString *keyPath, id oldValue, id newValue) {
        NSLog(@"---oldValue: %@, newValue: %@", oldValue, newValue);
    }];
}

- (IBAction)removeObserver:(id)sender {
    [self.textField zj_removeObserve:self forKeyPath:@"text"];
}

- (void)textFieldTextDidChange {
    self.textField.text = self.textField.text;
}

- (void)dealloc {
    [self removeObserver:nil];
}

@end

File: 自写/Demo/ObjectiveCRuntimeDemo/ObjectiveCRutimeDemo/KVO/KVOController.swift
//
//  KVOController.swift
//  ObjectiveCRutimeDemo
//
//  Created by luozhijun on 2017/3/17.
//  Copyright © 2017年 RickLuo. All rights reserved.
//

import UIKit

class KVOController: UIViewController {

    @IBOutlet var textField: UITextField!
    
    override func viewDidLoad() {
        NotificationCenter.default.addObserver(self, selector: #selector(textFieldTextChanged), name: Notification.Name.UITextFieldTextDidChange, object: nil)
    }
    
    func textFieldTextChanged() {
        textField.text = textField.text
    }
    
    @IBAction func observe(_ sender: Any) {
        textField.zj_addObserver(self, forKeyPath: "text") { (observer, keyPath, oldValue, newValue) in
            print("oldValue: \(oldValue), newValue: \(newValue)")
        }
        print(textField.classForCoder)
        print(textField.superclass!)
    }
    
    @IBAction func cancelObservation(_ sender: Any) {
        textField.zj_removeObserver(self, forKeyPath: "text")
    }
}
